### ISSUE TITLE
Have coordinate transformations preserve (float?) dtypes

### ISSUE THREAD

**Issue by eteq** [astropy/astropy#8870]
This *may* be more a part of `units`/`Quantity` than `coordinates`, but the proximate problem is coordinates so I'm starting there... the core problem is this:

```
>>> import numpy as np
>>> from astropy import units as u
>>> from astropy import coordinates as coo
>>> icrs = coo.ICRS(coo.CartesianRepresentation(np.array([1,2,3], dtype='float32')*u.kpc))
>>> icrs.data.xyz.dtype
dtype('float32')
>>> icrs.transform_to(coo.Galactic).data.xyz.dtype
dtype('float64')
```

that is, if I start with a dtype that's `float32` the  transformation process turns it into `float64`. It's not clear we formally expected it to stay the same, but it would be *useful* for some applications of the dtype is preserved...

**Comment by adrn** [comment]
This *almost* feels like a bug.

**Comment by pllim** [comment]
Why does this sound so familiar... #8614 and #8613 ?

**Comment by mhvk** [comment]
@eteq - this is because the `erfa` routines used to transform only take in doubles. The way ufuncs work, `float32` gets automatically converted to `double` and then one gets `double` out.

In principle, one could create `float32` wrappers around the erfa ufuncs, which especially for representations is not unreasonable (see https://github.com/astropy/astropy/issues/8615#issuecomment-495666913; the issue has a bit more general thought as well).

This would help if memory is the main worry, but if the goal is a speed-up, then one might be better off writing specific `float` C routines for the coordinate transformation.

I think the above is not a bad idea for `Representations`, which I think of as 3-vectors and which could reasonably have different dtype, but I'm still a bit weary of doing it for coordinate transformations generally.

**Comment by eteq** [comment]
@mhvk - that makes sense, thanks for clarifying.  I was misled into thinking it was different because I was *actually* trying this for the dtype `numpy.float128`,and it *failed* (with the same error as in #8615) until I re-compiled on a system that actually had an appropriate C-level `long double`, but I think that might have been the numpy 1.16 problem coming back to mislead me... 

I should elaborate the use case a bit more: I actually am more inerested in preserving `float128` because that would be a way to improve precision of coordinate differences marginally in some cases . I was thinking `float32` is more portable since there's no awkward "float128 isn't really 128 bits" issue (numpy/numpy#10288) problem, but perhaps the 128 use case is clearer because in that case 64 is a *demotion* rather than *promotion* (as in the 32 case).  So it's easier to justify why it's worth the trouble for the transformations.

It's a pretty non-trivial change erfa change to do this, but what do you think about this: we *could* imagine re-working the erfa generator to produce wrappers for `float`, `double`, and `long double`, and match the numpy types appropriately to those cases. A cursory glance at the generator makes me think that's mechanically not super-complicated, but I am not sure if I'm glossing over some of the more subtle numpy dtype issues that might arise here (especially given that it might exacerbate the "compiler does not match system on which the erfa c code was generated on" problem that bit us in the 3.2 release...).  What do you think, @mhvk?

**Comment by mhvk** [comment]
I agree that for some erfa routines it really should not be difficult - and that particularly includes the transformation routines we need for the representations. I think we'd want to use particular numpy types rather than C types (`npy_double`), etc., but that is really an implementation detail - it will just set us up more easily for when numpy does get a proper `float128`...

p.s. Of course, for many other erfa routines, like the time conversion ones, it makes no sense, as those really are written for doubles.

**Comment by mhvk** [comment]
p.s. It actually troubles me that the following silently works:
```
from astropy._erfa.ufunc import s2c 
s2c(np.array(1, dtype='f16'), np.array(1., dtype='f16'))
# array([0.29192658, 0.45464871, 0.84147098])
```

**Comment by eteq** [comment]
> It actually troubles me that the following silently works

You mean because you expected it to fail, or non-silently work, @mhvk ?

**Comment by mhvk** [comment]
Either, really, but it is a generally true for ufuncs that they allow "same_kind" casting. In principle, we could avoid this by adding `casting='safe'`, though given that we do not have full precision routines, it seems rather pointless.
